File: assets/js/hooks/gap_breath_timeline.js
import * as d3 from "d3"

const HEIGHT = 170
const MARGIN = { top: 18, right: 14, bottom: 22, left: 14 }
const MIN_BLOB_WIDTH = 4
const MAX_HALF_HEIGHT = 46
const MIN_HALF_HEIGHT = 1.2
const BASELINE_COLOR = "rgba(255,255,255,0.08)"

export const GapBreathTimeline = {
  mounted() {
    this.data = null

    this.handleEvent("update-gaps", (data) => {
      this.data = data
      this.render()
    })

    this._onResize = () => this.render()
    window.addEventListener("resize", this._onResize)
  },

  destroyed() {
    window.removeEventListener("resize", this._onResize)
    if (this.tooltip) this.tooltip.remove()
  },

  render() {
    if (!this.data) return
    const { transitions, imputedMemberships, dateRange, clusterColors, clusterNames } = this.data
    if (!dateRange || !transitions) return

    this.el.innerHTML = ""
    this.el.style.position = "relative"

    const width = this.el.clientWidth || 900
    const innerW = width - MARGIN.left - MARGIN.right
    const innerH = HEIGHT - MARGIN.top - MARGIN.bottom
    const midY = innerH / 2

    const t0 = parseDate(dateRange.start)
    const t1 = parseDate(dateRange.end)
    const x = d3.scaleTime().domain([t0, t1]).range([0, innerW])

    const svg = d3.select(this.el)
      .append("svg")
      .attr("width", width)
      .attr("height", HEIGHT)
      .style("display", "block")

    const g = svg.append("g")
      .attr("transform", `translate(${MARGIN.left},${MARGIN.top})`)

    g.append("line")
      .attr("x1", 0).attr("x2", innerW)
      .attr("y1", midY).attr("y2", midY)
      .attr("stroke", BASELINE_COLOR)
      .attr("stroke-width", 1)

    const axis = d3.axisBottom(x)
      .ticks(d3.timeYear.every(1))
      .tickFormat(d3.timeFormat("%Y"))
      .tickSize(0)
      .tickPadding(8)

    g.append("g")
      .attr("transform", `translate(0,${innerH + 2})`)
      .call(axis)
      .call(sel => sel.select(".domain").remove())
      .call(sel => sel.selectAll("text")
        .attr("fill", "#666")
        .attr("font-size", "10px"))

    const blobs = transitions
      .map(t => buildBlob(t, imputedMemberships, clusterColors, clusterNames || {}))
      .filter(Boolean)

    if (blobs.length === 0) {
      g.append("text")
        .attr("x", innerW / 2)
        .attr("y", midY)
        .attr("text-anchor", "middle")
        .attr("fill", "#666")
        .attr("font-size", "11px")
        .text("no gaps in range")
      return
    }

    const hook = this
    const area = d3.area()
      .curve(d3.curveCatmullRom.alpha(0.5))
      .x(d => x(d.date))
      .y0(d => midY - d.half)
      .y1(d => midY + d.half)

    g.append("g")
      .selectAll("path.breath")
      .data(blobs)
      .join("path")
      .attr("class", "breath")
      .attr("d", d => {
        const widthPx = x(d.endDate) - x(d.startDate)
        if (d.samples.length >= 2 && widthPx >= MIN_BLOB_WIDTH) {
          return area(d.samples)
        }
        const cx = x(d.startDate)
        const r = Math.max(MIN_BLOB_WIDTH / 2, widthPx / 2)
        return `M ${cx - r} ${midY - d.peakHalf} L ${cx + r} ${midY - d.peakHalf} L ${cx + r} ${midY + d.peakHalf} L ${cx - r} ${midY + d.peakHalf} Z`
      })
      .attr("fill", d => d.color)
      .attr("fill-opacity", 0.72)
      .attr("stroke", d => d.color)
      .attr("stroke-opacity", 0.9)
      .attr("stroke-width", 0.6)
      .style("cursor", "pointer")
      .on("mouseenter", function(event, d) {
        d3.select(this).attr("fill-opacity", 1)
        hook.showTooltip(event, d)
      })
      .on("mousemove", function(event, d) { hook.moveTooltip(event) })
      .on("mouseleave", function() {
        d3.select(this).attr("fill-opacity", 0.72)
        hook.hideTooltip()
      })
      .on("click", function(event, d) {
        hook.pushEvent("gap_selected", { start: d.startStr, length: d.length })
      })

    this.tooltip = d3.select(this.el)
      .append("div")
      .attr("class", "breath-tooltip")
      .style("position", "absolute")
      .style("pointer-events", "none")
      .style("background", "rgba(10,10,26,0.94)")
      .style("color", "#eee")
      .style("padding", "8px 12px")
      .style("border-radius", "6px")
      .style("border", "1px solid rgba(255,255,255,0.08)")
      .style("font-size", "11px")
      .style("line-height", "1.5")
      .style("white-space", "nowrap")
      .style("z-index", "100")
      .style("opacity", 0)
  },

  showTooltip(event, d) {
    const plural = d.length === 1 ? "day" : "days"
    const html = `
      <strong>${d.startStr}</strong> · ${d.length} ${plural}
      <div style="color:${d.color}">${d.clusterName}</div>
    `
    this.tooltip.html(html).style("opacity", 1)
    this.moveTooltip(event)
  },

  moveTooltip(event) {
    const rect = this.el.getBoundingClientRect()
    const top = event.clientY - rect.top + 14
    const left = Math.min(event.clientX - rect.left + 14, rect.width - 180)
    this.tooltip.style("top", top + "px").style("left", left + "px")
  },

  hideTooltip() {
    if (this.tooltip) this.tooltip.style("opacity", 0)
  }
}

function parseDate(str) {
  return d3.timeParse("%Y-%m-%d")(str)
}

function addDays(date, n) {
  const d = new Date(date)
  d.setDate(d.getDate() + n)
  return d
}

function fmtDate(date) {
  return d3.timeFormat("%Y-%m-%d")(date)
}

function buildBlob(transition, imputed, clusterColors, clusterNames) {
  const gap = transition.gap || {}
  const startStr = gap.start
  const length = gap.length
  if (!startStr || !length) return null

  const startDate = parseDate(startStr)
  if (!startDate) return null

  const totals = {}
  const samples = []
  let peakHalf = MIN_HALF_HEIGHT

  for (let i = 0; i < length; i++) {
    const day = addDays(startDate, i)
    const key = fmtDate(day)
    const mems = (imputed && imputed[key]) || transition.before || {}
    const { cluster, strength } = dominantOf(mems)
    if (cluster) totals[cluster] = (totals[cluster] || 0) + strength
    const half = Math.max(MIN_HALF_HEIGHT, strength * MAX_HALF_HEIGHT)
    if (half > peakHalf) peakHalf = half
    samples.push({ date: day, half, strength, cluster })
  }

  const dominantCluster = Object.entries(totals)
    .sort((a, b) => b[1] - a[1])[0]?.[0]
  const color = (dominantCluster && clusterColors[dominantCluster]) || "#888"
  const clusterName = (dominantCluster && clusterNames[dominantCluster]) || dominantCluster || "—"

  const endDate = addDays(startDate, Math.max(length - 1, 0))

  return {
    startStr,
    startDate,
    endDate,
    length,
    samples,
    peakHalf,
    color,
    clusterName,
  }
}

function dominantOf(mems) {
  if (!mems) return { cluster: null, strength: 0 }
  let bestK = null
  let bestV = 0
  for (const k in mems) {
    const v = mems[k]
    if (typeof v === "number" && v > bestV) {
      bestV = v
      bestK = k
    }
  }
  return { cluster: bestK, strength: bestV }
}

File: lib/fugue_web/live/mood_live/gap_breath_timeline.ex
defmodule FugueWeb.MoodLive.GapBreathTimeline do
  @moduledoc """
  Breath-shaped visualization of silent stretches (data gaps). Width = duration,
  vertical half-height = model confidence, fill = dominant imputed state.
  Replaces the former `GapBreathTimeline` JS hook.
  """

  use Phoenix.Component

  alias FugueWeb.MoodLive.SvgMath

  @svg_width 900
  @height 170
  @m_top 18
  @m_right 14
  @m_bottom 22
  @m_left 14
  @inner_w @svg_width - @m_left - @m_right
  @inner_h @height - @m_top - @m_bottom
  @mid_y @inner_h / 2
  @min_blob_width 4
  @max_half_height 46
  @min_half_height 1.2

  attr :transitions, :list, default: []
  attr :imputed_memberships, :map, default: %{}
  attr :date_range, :any, default: nil
  attr :cluster_colors, :map, default: %{}
  attr :cluster_names, :map, default: %{}

  def timeline(assigns) do
    case assigns.date_range do
      nil ->
        assign(assigns, :empty?, true) |> render_empty()

      %{start: start_str, end: end_str} ->
        min_d = Date.from_iso8601!(start_str)
        max_d = Date.from_iso8601!(end_str)
        span = max(Date.diff(max_d, min_d), 1)

        x_fn = fn date_str ->
          Date.diff(Date.from_iso8601!(date_str), min_d) / span * @inner_w
        end

        blobs =
          assigns.transitions
          |> Enum.map(fn t ->
            build_blob(
              t,
              assigns.imputed_memberships,
              assigns.cluster_colors,
              assigns.cluster_names,
              x_fn
            )
          end)
          |> Enum.reject(&is_nil/1)

        year_ticks =
          for y <- min_d.year..max_d.year,
              jan1 = Date.new!(y, 1, 1),
              Date.compare(jan1, min_d) != :lt and Date.compare(jan1, max_d) != :gt do
            %{x: x_fn.(Date.to_iso8601(jan1)), label: Integer.to_string(y)}
          end

        assigns =
          assign(assigns,
            blobs: blobs,
            year_ticks: year_ticks,
            empty?: blobs == [],
            svg_width: @svg_width,
            svg_height: @height,
            inner_w: @inner_w,
            inner_h: @inner_h,
            mid_y: @mid_y,
            g_transform: "translate(#{@m_left},#{@m_top})",
            axis_transform: "translate(0,#{@inner_h + 2})"
          )

        render_full(assigns)
    end
  end

  defp render_empty(assigns) do
    ~H"""
    <div id="gap-breath-timeline"></div>
    """
  end

  defp render_full(assigns) do
    ~H"""
    <div id="gap-breath-timeline" phx-hook="HoverTooltip" style="position: relative;">
      <svg
        viewBox={"0 0 #{@svg_width} #{@svg_height}"}
        preserveAspectRatio="xMidYMid meet"
        style="display: block; width: 100%;"
      >
        <g transform={@g_transform}>
          <line
            x1="0"
            x2={@inner_w}
            y1={SvgMath.fmt(@mid_y)}
            y2={SvgMath.fmt(@mid_y)}
            stroke="rgba(255,255,255,0.08)"
            stroke-width="1"
          />

          <g transform={@axis_transform}>
            <%= for t <- @year_ticks do %>
              <text
                x={SvgMath.fmt(t.x)}
                y="12"
                text-anchor="middle"
                fill="#666"
                font-size="10px"
              >
                {t.label}
              </text>
            <% end %>
          </g>

          <%= if @empty? do %>
            <text
              x={SvgMath.fmt(@inner_w / 2)}
              y={SvgMath.fmt(@mid_y)}
              text-anchor="middle"
              fill="#666"
              font-size="11px"
            >
              no gaps in range
            </text>
          <% else %>
            <%= for blob <- @blobs do %>
              <path
                class="breath"
                d={blob.d}
                fill={blob.color}
                fill-opacity="0.72"
                stroke={blob.color}
                stroke-opacity="0.9"
                stroke-width="0.6"
                style="cursor: pointer;"
                data-tooltip={blob.tooltip}
                phx-click="gap_selected"
                phx-value-start={blob.start_str}
                phx-value-length={blob.length}
              />
            <% end %>
          <% end %>
        </g>
      </svg>

      <style>
        .breath { transition: fill-opacity 0.15s; }
        .breath:hover { fill-opacity: 1 !important; }
      </style>
    </div>
    """
  end

  # --- Blob construction ---

  defp build_blob(transition, imputed, cluster_colors, cluster_names, x_fn) do
    gap = transition["gap"] || %{}
    start_str = gap["start"]
    length = gap["length"]

    with true <- is_binary(start_str) and is_integer(length) and length > 0,
         {:ok, start_date} <- Date.from_iso8601(start_str) do
      {samples, totals, peak_half} = walk_gap(start_date, length, imputed, transition)

      dominant = totals |> Enum.max_by(fn {_, v} -> v end, fn -> {nil, 0} end) |> elem(0)
      color = (dominant && Map.get(cluster_colors, dominant)) || "#888"
      cluster_name = (dominant && Map.get(cluster_names, dominant, dominant)) || "--"

      end_date = Date.add(start_date, max(length - 1, 0))
      end_str = Date.to_iso8601(end_date)
      start_x = x_fn.(start_str)
      end_x = x_fn.(end_str)
      width_px = end_x - start_x

      d =
        if length(samples) >= 2 and width_px >= @min_blob_width do
          upper = Enum.map(samples, fn s -> {x_fn.(s.date), @mid_y - s.half} end)
          lower = Enum.map(samples, fn s -> {x_fn.(s.date), @mid_y + s.half} end)
          area_path(upper, lower)
        else
          cx = start_x
          r = max(@min_blob_width / 2, width_px / 2)
          top = @mid_y - peak_half
          bot = @mid_y + peak_half

          "M#{SvgMath.fmt(cx - r)} #{SvgMath.fmt(top)} " <>
            "L#{SvgMath.fmt(cx + r)} #{SvgMath.fmt(top)} " <>
            "L#{SvgMath.fmt(cx + r)} #{SvgMath.fmt(bot)} " <>
            "L#{SvgMath.fmt(cx - r)} #{SvgMath.fmt(bot)} Z"
        end

      plural = if length == 1, do: "day", else: "days"

      tooltip =
        ~s|<strong>#{start_str}</strong> · #{length} #{plural}<div style="color:#{color}">#{cluster_name}</div>|

      %{
        start_str: start_str,
        length: length,
        color: color,
        d: d,
        tooltip: tooltip
      }
    else
      _ -> nil
    end
  end

  defp walk_gap(start_date, length, imputed, transition) do
    Enum.reduce(0..(length - 1), {[], %{}, @min_half_height}, fn i,
                                                                 {samples_acc, totals_acc,
                                                                  peak_acc} ->
      day = Date.add(start_date, i)
      key = Date.to_iso8601(day)
      mems = Map.get(imputed || %{}, key) || Map.get(transition, "before", %{}) || %{}
      {cluster, strength} = dominant_of(mems)

      new_totals =
        if cluster,
          do: Map.update(totals_acc, cluster, strength, &(&1 + strength)),
          else: totals_acc

      half = max(@min_half_height, strength * @max_half_height)
      new_peak = max(peak_acc, half)
      sample = %{date: key, half: half, strength: strength, cluster: cluster}
      {[sample | samples_acc], new_totals, new_peak}
    end)
    |> then(fn {samples, totals, peak} -> {Enum.reverse(samples), totals, peak} end)
  end

  defp dominant_of(mems) when is_map(mems) and map_size(mems) > 0 do
    Enum.reduce(mems, {nil, 0}, fn {k, v}, {best_k, best_v} ->
      if is_number(v) and v > best_v, do: {k, v}, else: {best_k, best_v}
    end)
  end

  defp dominant_of(_), do: {nil, 0}

  defp area_path(upper, lower) do
    upper_path = SvgMath.basis_path(upper)
    lower_path = SvgMath.basis_path(Enum.reverse(lower))

    lower_continued =
      case lower_path do
        "M" <> rest -> "L" <> rest
        "" -> ""
        other -> other
      end

    upper_path <> lower_continued <> "Z"
  end
end
